### PR TITLE
Add DeltaBuild into Arcade

### PR DESCRIPTION
**Pull request by tekian**
### To double check:

* [ ] The right tests are in and and the right validation has happened.  Guidance:  https://github.com/dotnet/arcade/tree/main/Documentation/Validation

### CODE REVIEW

File: Arcade.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29019.234
+# Visual Studio Version 17
+VisualStudioVersion = 17.5.33502.453
```
Comment (tekian) [inline Arcade.sln:4]: Need to revert.

**Comment by markwilkie** [comment]
If @riarenas , @chcosta , @AlitzelMendez  could take a look that'd be super helpful.  Jan and I have been talking, and basically he'd like to add this builddelta tool to Arcade.  It's not intended to be in the primary tool chain path, but available for those who want it.

**Comment by riarenas** [comment]
I lack the necessary context, so I'll ask:

* What is the business goal of having this tool available through arcade instead of however it's distributed now?
* Who is going to own the tool, including paying attention to issues related to it? Especially after the recent reorg I don't think we should pick up anything new for now, even if we would just be directing any issues to another team.
* The PR is still a draft, I assume there will be tests coming in future iterations?

File: src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    <RootNamespace>DeltaBuild</RootNamespace>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="LibGit2Sharp" Version="$(LibGit2SharpVersion)" />
+    <PackageReference Include="MSBuild.StructuredLogger" Version="$(MSBuildStructuredLoggerVersion)" />
+    <PackageReference Include="McMaster.Extensions.CommandLineUtils" Version="$(McMasterExtensionsCommandLineUtils)" />
+  </ItemGroup>
+
+  <Import Project="$(RepoRoot)eng\BuildTask.targets" />
```
Comment (riarenas) [inline src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:20]: This looks to just be a tool and not a build task. I don't think this is needed?

Comment (riarenas) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:20]: [AsmDiff](https://github.com/dotnet/arcade/tree/b8007eed82adabd50c604a9849277a6e7be5c971/src/Microsoft.DotNet.AsmDiff) should be an example for another dotnet tool that is not a build task.

**Comment by RussKie** [comment]
I plan to start reviewing this early next week.

**Comment by tekian** [comment]
> What is the business goal of having this tool available through arcade instead of however it's distributed now?
We'd like to reuse the tool between multiple .NET repositories. Today, the tool is not distributed in any way. Source of the tool is built as a prerequisite before source of unrelated project is built. It's a new tool.

> Who is going to own the tool, including paying attention to issues related to it?
That's going to be myself, @mobratil, @RussKie and more broadly the .NET teams using it.

> Are tests coming in future iterations?
Absolutely. The tool works as it is now, but we wouldn't be able to catch any regressions if we decide to add or change the behavior. After we have tool published as a .NET tool, I'm going to add a test suite for it.

File: src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj (hunk 2 above)
Comment (tekian) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj]: Thanks for pointing this out. I wasn't sure what is it used for, some tools had it. It's now removed and I done couple of additional steps to alins with `AsmDiff`, namely:

- Aligned namespaces to be `Microsoft.DotNet.DeltaBuild`.
- Added `ExcludeFromSourceBuild` to `true` and `StrongNameKeyId` to `Open` based on `AsmDiff`.

**Comment by RussKie** [comment]
image

You're setting yourself for a lot of pain.... :)

**Comment by RussKie** [comment]
The Linux build fails with:
```
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm/native/libgit2-b7bad55.so
##[error].packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : (NETCORE_ENGINEERING_TELEMETRY=Publish) Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm/native/libgit2-b7bad55.so
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm/native/libgit2-b7bad55.so
##[error].packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : (NETCORE_ENGINEERING_TELEMETRY=Publish) Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm/native/libgit2-b7bad55.so
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm64/native/libgit2-b7bad55.so
##[error].packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : (NETCORE_ENGINEERING_TELEMETRY=Publish) Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm64/native/libgit2-b7bad55.so
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm64/native/libgit2-b7bad55.so
##[error].packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : (NETCORE_ENGINEERING_TELEMETRY=Publish) Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-arm64/native/libgit2-b7bad55.so
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-x64/native/libgit2-b7bad55.so
##[error].packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : (NETCORE_ENGINEERING_TELEMETRY=Publish) Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-x64/native/libgit2-b7bad55.so
/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-x64/native/libgit2-b7bad55.so
##[error].packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23164.2/tools/Publish.proj(242,5): error : (NETCORE_ENGINEERING_TELEMETRY=Publish) Invalid ELF BuildID '<null>' for tools/net8.0/any/runtimes/linux-musl-x64/native/libgit2-b7bad55.so
  Symbol publish finished
```

https://github.com/dotnet/symstore/issues/81 seems related, and the solution is to exclude libgit2-b7bad55.so from the symbol packages...

**Comment by tekian** [comment]
@riarenas With @RussKie 's help, we now have all 15 checks green. What should be our next step? :-)

File: start-vs.cmd [hunk 3]
```diff
@@ -0,0 +1,42 @@
+@echo off
```
Comment (riarenas) [inline start-vs.cmd:1]: The addition of this file does not seem related to deltabuild

**Comment by riarenas** [comment]
I'd like to see an official test build with these changes, especially due to the workarounds for symbol publishing errors you are introducing that will only be exercised during official builds.

@RussKie has done this before, and the steps are basically to follow the validation steps in this document: https://github.com/dotnet/arcade/blob/main/Documentation/CorePackages/Publishing.md#validating-the-changes

File: start-vs.cmd (hunk 3 above)
Comment (RussKie) [reply start-vs.cmd:1]: This script makes possible to open the solution in Visual Studio.

Comment (riarenas) [reply start-vs.cmd:1]: My sense is that this script should not live in the root of the repo or it will give the idea that it's supported by the Arcade team. Maybe move it to the tool's directory?

Comment (RussKie) [reply start-vs.cmd:1]: That's the script to open Arcade.sln, and it should live at the root. The same as in dotnet/winforms, dotnet/aspnetcore, dotnet/r9, etc....

Comment (riarenas) [reply start-vs.cmd:1]: It makes me uncomfortable to add tooling to the repo root as part of a PR to add an optional tool.

Comment (riarenas) [reply start-vs.cmd:1]: That said, if @markwilkie is fine with it I don't mind having in there.

Comment (RussKie) [reply start-vs.cmd:1]: I can submit a separate PR for the script, though it feels kind of waste of compute :) Your call.

Comment (markwilkie) [reply start-vs.cmd:1]: I too feel (generally) weird about adding something to the root.  Thoughts though @mmitche ?

Comment (mmitche) [reply start-vs.cmd:1]: This is a point in time problem, too, I think. I think it's only required until we insert .NET 8 into VS? I've never seen this problem on servicing branches when building arcade, or any point after the SDK is included in VS.

Comment (RussKie) [reply start-vs.cmd:1]: @mmitche @riarenas I'm genuinely curious how are you building solutions in VS that target the dev version of .NET SDK (which is currently .NET 8). It's always been a pain across all .NET repos, that's why scripts such as this were added to restore the toolset, configure the right variable and open a solution in VS.

File: src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    <ToolCommandName>dotnet-deltabuild</ToolCommandName>
+    <ExcludeFromSourceBuild>true</ExcludeFromSourceBuild>
+    <StrongNameKeyId>Open</StrongNameKeyId>
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="LibGit2Sharp" Version="$(LibGit2SharpVersion)" />
+    <PackageReference Include="MSBuild.StructuredLogger" Version="$(MSBuildStructuredLoggerVersion)" />
+    <PackageReference Include="McMaster.Extensions.CommandLineUtils" Version="$(McMasterExtensionsCommandLineUtils)" />
```
[the remark this answers is not in the document]
Comment (RussKie) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:18]: It's already used in few places: https://github.com/dotnet/arcade/search?q=McMaster

[the remark this answers is not in the document]
Comment (tekian) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:18]: Actually, we had `System.CommandLine` in DeltaBuild before creating this PR into Arcade. But then I've noticed that `System.CommandLine` is not used within Arcade's suite of projects: https://github.com/dotnet/arcade/search?q=McMaster. So I have looked around and saw that, as @RussKie pointed out, few projects use McMaster's command line parsing. So I've switched to that. 😉

[the remark this answers is not in the document]
Comment (tekian) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:18]: The thing is System.CommandLine has only preview versions. Won't that cause problems? I'm happy to move either way, as I've said, we had System.CommandLine originally. But at the same time, maybe such an upgrade from McMaster to System.CommandLine should be done on all the tooling in a separate PR?

[the remark this answers is not in the document]
Comment (mmitche) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:18]: It only has preview versions, but is used by all of our shipped tooling at this point (e.g. the dotnet SDK). Do you happen to have the original System.CommandLine code you wrote before replacing it?

[the remark this answers is not in the document]
Comment (tekian) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:18]: Yep. I'll check that in instead of McMaster.

File: src/Microsoft.DotNet.DeltaBuild/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+Optionally, you may provide additional information such as:
+- `-d|--debug` if you'd like to see additional information.
+- `-o|--output-json` to capture result in a file.
+- `-bbl|--branch-binlog` the binlog for the target branch to account for changes where file or project may have been deleted.
+- `-b|--branch` to specify target branch (default: `origin/main`)
+
+
+## Results
+
+Depending on size of the solution and relative change, measured speedup can vary from 25% to 85%.
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.DeltaBuild/README.md:55]: cc @rainersigwald @dsplaisted

File: src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (tekian) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj]: Switched back to `System.CommandLine`.

**Review by RussKie — APPROVED** [review]
👍 
Few suggestions to the readme.

**Comment by RussKie** [comment]
@mmitche @riarenas  I pushed this branch to https://dnceng.visualstudio.com/internal/_git/dotnet-arcade?version=GBdeltabuild. I also created a channel:
```
(4877) https://dev.azure.com/dnceng/internal/_git/dotnet-arcade @ deltabuild -> General Testing
```
The [build](https://dev.azure.com/dnceng/public/_artifacts/feed/general-testing) appears to have [published the artifacts to the BAR](https://dev.azure.com/dnceng/internal/_build/results?buildId=2209094&view=logs&j=d17d50ef-afac-5d4b-95a6-3014ce5cf7a4&t=95fe4740-169e-5ca1-2f88-330db40deac4&l=74):
```
  Metadata has been pushed. Build id in the Build Asset Registry is '183022'
  Found the following default channels:
      https://dev.azure.com/dnceng/internal/_git/dotnet-arcade@deltabuild => (529) General Testing
  Determined build will be added to the following channels: [529]
```
...but those [don't appear to have been pushed to the NuGet feed](https://dev.azure.com/dnceng/internal/_build/results?buildId=2209094&view=logs&j=226748d0-f812-5437-d3f0-2dd291f5666e&t=bad11196-972e-5d03-45a8-9db526506073&l=28).

What am I missing?

**Comment by RussKie** [comment]
Also, more importantly, what else we need to do to get this merged?

**Comment by mmitche** [comment]
It's because the shas match between the public and internal repos, so Maestro will preference the public repo URI. In that case the default channel doesn't apply.

I very rarely recommend using default channels for one-off testing like this. Instead, just do the following after the build is complete:

`darc add-build-to-channel --id <BAR ID> --channel "General Testing"`

and wait for it to complete.

**Comment by RussKie** [comment]
Right, thanks. I was kind of wondering whether the docs should have said "General Testing **Internal**". I think the docs should be updated, though there are three possibilities:
- Option 1: Instead of "_Create a branch on the Azure DevOps internal mirror of the repo that includes the pipeline changes._" use the public repo. Then "General Testing" would work.
- Option 2: Instead of "General Testing" specify "General Testing Internal".
- Option 3: Remove default channel creation steps and use `darc add-build-to-channel --id <BAR ID> --channel "General Testing"` instead.

Which version do you prefer? I'm happy to update the docs.

**Comment by RussKie** [comment]
Available test artifacts:
* https://dev.azure.com/dnceng/public/_artifacts/feed/general-testing/NuGet/Microsoft.DotNet.DeltaBuild/overview/8.0.0-beta.23325.1

File: src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj [hunk 6]
```diff
@@ -0,0 +1,37 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>$(NetCurrent)</TargetFramework>
```
Comment (RussKie) [inline src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:5]: By targeting the current TFM only we make the tool only usable by projects targeting .NET 8. Which isn't good. We need to be targeting something like net472.

**Review by RussKie — COMMENTED** [review]
We need a way to reference and invoke the tool.

I'm not exactly clear on how to do the first step - i.e., reference. The tool has to run (if opted in) before the build because it defines the subset of projects that get built.
One way could be via global.json `msbuild-sdks` node, e.g.:
```diff
  "msbuild-sdks": {
    "Microsoft.DotNet.Arcade.Sdk": "8.0.0-beta.23322.2",
    "Microsoft.DotNet.Helix.Sdk": "8.0.0-beta.23322.2",
+   "Microsoft.DotNet.DeltaBuild": "8.0.0-beta.23325.1"
  }
```
...but it means we need to create DeltaBuildSdk.  I don't know how this works though...

The other option is to add an unconditional reference to https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.Arcade.Sdk/tools/Tools.proj, e.g.:
```diff
  <ItemGroup>
    <PackageReference Include="Microsoft.DotNet.NuGetRepack.Tasks" Version="$(MicrosoftDotnetNuGetRepackTasksVersion)" Condition="'$(UsingToolNuGetRepack)' == 'true'" IsImplicitlyDefined="true" />
+   <PackageReference Include="Microsoft.DotNet.DeltaBuild" Version="8.0.0-beta.23325.1" IsImplicitlyDefined="true">
+     <PrivateAssets>all</PrivateAssets>
+   </PackageReference>
  </ItemGroup>
```
...and then create DeltaBuild.props and DeltaBuild.targets that contain some targets that run before `Build` (or `Restore`?) target. Again this requires further thoughts and iterations.

**Comment by RussKie** [comment]
@markwilkie  @mmitche @riarenas is there someone we could we talk to learn how to best plug this tool into the Arcade SDK and the whole build process?

**Comment by mmitche** [comment]
> @markwilkie @mmitche @riarenas is there someone we could we talk to learn how to best plug this tool into the Arcade SDK and the whole build process?

I think the package ref in tools.proj, conditionalized based on whether the feature is enabled (this is important since it shouldn't be enabled by default) is a good way to start.

**Comment by RussKie** [comment]
> > @markwilkie
> >             
> >             Mark Wilkie
> >             FTE @mmitche
> >             
> >             Matt Mitchell (.NET)
> >             FTE @riarenas
> >             
> >             Ricardo Arenas
> >             FTE is there someone we could we talk to learn how to best plug this tool into the Arcade SDK and the whole build process?
> 
> I think the package ref in tools.proj, conditionalized based on whether the feature is enabled (this is important since it shouldn't be enabled by default) is a good way to start.

I tried to do this locally by defining `UseDeltaBuild=true` property in a test project's Directory.Build.props and conditionalising the import in the Arcade's Tools.proj (in my local nuget cache), and it didn't work. The binlog was showing the property is set, however, that was too late for the Tools.proj eval, or it had no knowledge of that...

**Comment by mmitche** [comment]
> Right, thanks. I was kind of wondering whether the docs should have said "General Testing **Internal**". I think the docs should be updated, though there are three possibilities:
> 
> * Option 1: Instead of "_Create a branch on the Azure DevOps internal mirror of the repo that includes the pipeline changes._" use the public repo. Then "General Testing" would work.
> * Option 2: Instead of "General Testing" specify "General Testing Internal".
> * Option 3: Remove default channel creation steps and use `darc add-build-to-channel --id <BAR ID> --channel "General Testing"` instead.
> 
> Which version do you prefer? I'm happy to update the docs.

The internal vs. public channel isn't really about the repo being public or internal. It's about whether the artifacts **can** be public. For instance, we publish wpf-int to public channels, even though the source is internal. It's simply that Maestro has some logic in it that preferences the repo URI to be public, if the commit built is public. So it takes the AzDO URI + sha that was built officially, e.g. dotnet-extensions @ foosha, translates the URI to what the github equivalent would be, i.e. dotnet/extensions @ foosha, then attempts to do an anonymous get. If the URI 404's, then it says that the URI of the repo built was the original URI, otherwise it adds github.

Default channels always preference a build's github uri over the azdo one rather than letting both go through so that we cannot **mistakenly** publish internal bits to public endpoints. A good example of how this could happen is this:
- A dev has a security fix that they open in an AzDO PR, mistakenly pointing the target branch to release/5.0.
- It merges and officially builds.
- The default channel picks up the release/5.0 branch + AzDO URI and publishes public bits.

Instead in this case, the default channel for public repos always pulls from GitHub, but the commit doesn't show up on github. So the AzDO URI does not match the default channel and no sensitive bits are accidentally published.

TLDR: It's really just about what the purpose of your bits are. You may assign any build to a channel based on whether you want the bits to be public or not. You need to take into account whether commit being built is public or not when you set up a default channel. However, it is simpler to just manually assign in these one off situations, because you may be pushing the commit you build to both github and azdo, or only azdo.

**Comment by RussKie** [comment]
Thank you for taking time to explain this in details. I'll open a PR proposing an update to the docs.

File: src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj (hunk 6 above)
Comment (tekian) [reply src/Microsoft.DotNet.DeltaBuild/Microsoft.DotNet.DeltaBuild.csproj:5]: I had .NET Framework TFM but there was a build warning/error that a .NET Tool can only be .NET Core. Should we add .NET 7? Ideas?

File: src/Microsoft.DotNet.DeltaBuild.Tests/Microsoft.DotNet.DeltaBuild.Tests.csproj [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>net7.0;$(NetCurrent)</TargetFrameworks>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.DeltaBuild.Tests/Microsoft.DotNet.DeltaBuild.Tests.csproj:4]: Arcade usually only target NetCurrent. Why do you need to support net7.0?

Comment (RussKie) [reply src/Microsoft.DotNet.DeltaBuild.Tests/Microsoft.DotNet.DeltaBuild.Tests.csproj:4]: This is a _dotnet tool_, which needs to work for all kinds of solutions that may target the whole plethora of TFMs. The Arcade SDK isn't exclusive to .NET 8 projects and, for example, can be used to build .NET Framework projects. 
https://github.com/microsoft/slngen tool is another example, it targets the whole specter of TFMs:
image

Comment (RussKie) [reply src/Microsoft.DotNet.DeltaBuild.Tests/Microsoft.DotNet.DeltaBuild.Tests.csproj:4]: Right now, an attempt to install this tool fails:
```
C:\>dotnet tool install -g Microsoft.DotNet.DeltaBuild --add-source https://pkgs.dev.azure.com/dnceng/public/_packaging/general-testing/nuget/v3/index.json --prerelease

C:\Users\jagutt\AppData\Local\Temp\79bb3973-33dd-43b6-b8e8-78303e953957\restore.csproj : error NU1202: Package Microsoft.DotNet.DeltaBuild 8.0.0-beta.23325.1 is not compatible with net7.0 (.NETCoreApp,Version=v7.0) / any. Package Microsoft.DotNet.DeltaBuild 8.0.0-b
eta.23325.1 supports: net8.0 (.NETCoreApp,Version=v8.0) / any
The tool package could not be restored.
```

Comment (ViktorHofer) [reply src/Microsoft.DotNet.DeltaBuild.Tests/Microsoft.DotNet.DeltaBuild.Tests.csproj:4]: But why do you need only net7.0 then? net6.0 is currently the minimum supported TFM and is represented as an msbuild property: `$(NetMinimum)`. net7.0 should probably also be replaced with `$(NetPrevious)`.
